RNA_folding.py

```diff
@@ -0,0 +1,319 @@
+
+    Returns: Dictionary of all possible stems with maximal stems as keys.
+    """
+
+    stem_dict = {}
+    n = bond_matrix.shape[0]
+
+    # Iterate through matrix looking for possible stems.
+    for i in range(n - (2 * min_stem + min_loop)):
+        for j in range(i + 2 * min_stem + min_loop - 1, n):
+            if bond_matrix[i, j]:
+                k = 1
+                while bond_matrix[i + k, j - k]:
+                    bond_matrix[i + k, j - k] = False
+                    k += 1
+
+                if k >= min_stem:
+                    # A 4-tuple is used to represent the stem.
+                    stem_dict[(i, i + k - 1, j - k + 1, j)] = []
+
+    # Iterate through all substems weakly contained in a maximal stem under inclusion.
+    for stem in stem_dict.keys():
+        stem_dict[stem].extend([(stem[0] + i, stem[0] + k, stem[3] - k, stem[3] - i)
+                                for i in range(stem[1] - stem[0] - min_stem + 2)
+                                for k in range(i + min_stem - 1, stem[1] - stem[0] + 1)])
+
+    return stem_dict
+
+
+def check_overlap(stem1, stem2):
+    """ Checks if 2 stems use any of the same nucleotides.
+
+    Args:
+        stem1: 4-tuple containing stem information.
+        stem2: 4-tuple containing stem information.
+
+    Returns: Boolean indicating if the two stems overlap.
+    """
+
+    if type(stem1) == str or type(stem2) == str:
+        return False
+
+    # Check if any endpoints of stem2 overlap with stem1.
+    for val in stem2:
+        if stem1[0] <= val <= stem1[1] or stem1[2] <= val <= stem1[3]:
+            return True
+    # Check if endpoints of stem1 overlap with stem2.
+    # Do not need to check all stem1 endpoints.
+    for val in stem1[1:3]:
+        if stem2[0] <= val <= stem2[1] or stem2[2] <= val <= stem2[3]:
+            return True
+
+    return False
+
+
+def pseudoknot_terms(stem_dict, min_stem=3, c=0.3):
+    """ Creates a dictionary with all possible pseudoknots as keys and appropriate penalty as as value pair.
+
+    The penalty is the parameter c times the product of the lengths of the two stems in the knot.
+
+    Args:
+        stem_dict: Dictionary with maximal stems as keys and list of weakly contained substems as values.
+        min_stem: Integer smallest number of consecutive bonds to be considered a stem.
+        c: Float parameter factor of the penalty on pseudoknots.
+
+    Returns: Dictionary with all possible pseudoknots as keys and appropriate penalty as as value pair.
+    """
+
+    pseudos = {}
+    # Look within all pairs of maximal stems for possible psuedoknots.
+    for stem1, stem2 in product(stem_dict.keys(), stem_dict.keys()):
+        # Using product instead of combinations allows for short asymmetric checks.
+        if stem1[0] + 2 * min_stem < stem2[1] and stem1[2] + 2 * min_stem < stem2[3]:
+            pseudos.update({(substem1, substem2): c * (substem1[1] - substem1[0]) * (substem2[1] - substem2[0])
```

Comment:
Do you need to add 1 when computing the stem lengths, like you do in other places?
```suggestion
            pseudos.update({(substem1, substem2): c * (substem1[1] - substem1[0] + 1) * (substem2[1] - substem2[0] + 1)
```

Reply:
Yeah... I suppose you're right. Ugly code, cleaner concept.